[spirv] Add OpControlBarrier and OpMemoryBarrier

denis0x0D opened the pull request:
This patch addresses https://github.com/tensorflow/mlir/issues/127 

ebnf:
`scope ::=` `"CrossDevice"` | `"Device"` | `"Workgroup"` | `"Subgroup"`
              | `"Invocation"`

    memory-semantics ::= `"None"` | `"Acquire"` | `"Release"`
                         | `"AcquireReleae"` | `"SequentiallyConsistent"`
                         | `"UniformMemory"`  | `"SubgroupMemory"`
                         | `"WorkgroupMemory"`  | `"CrossWorkgroupMemory"`
                         | `"AtomicCounterMemory"` | `"ImageMemory"`

    control-barrier-op ::= `spv.ControlBarrier` scope, scope, memory-semantics

    memory-barrier-op ::= `spv.MemoryBarrier` scope, memory-semantics




Example:
``` 
 spv.ControlBarrier "Workgroup", "Device", "Acquire|UniformMemory"
 spv.MemoryBarrier "Device", "Acquire|UniformMemory"

```
@antiagainst @MaheshRavishankar can you please, take a look on this?
Thanks!

MaheshRavishankar commented on include/mlir/Dialect/SPIRV/SPIRVBase.td:
```diff
@@ -227,6 +230,20 @@ class SPV_Optional<Type type> : Variadic<type>;
+// invocation. The storage classes specify to which subsets of memory these
+// constraints are to be applied. Storage classes not selected are not being
+// constrained.
+def SPV_MemorySemantics : IntOfWidths<[32]>;
```
Not sure about this approach. Should really be using an enum to describe all the valid semantics.. The enum would be an I32Attr (see other enum definitions in SPIRVBase.td). You could then let the operation take a list of these attributes (i.e. and ArrayAttr) and then having the instructions take an array of these

denis0x0D replied:
@MaheshRavishankar thanks for review, since that mask it would require to describe all possible combination of those bits  https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#Memory_Semantics_-id- to describe a type.
Also, regarding to the spec the operands must be a ssa-id of integer:
OpControlBarrier %int_2, %int_1, %int_2376 
it would require to creating of constants  while serizalizing.
If it's ok, I will update the patch.
Thanks.

antiagainst replied:
Technically this memory semantics `<id>` can be generated from another instruction like `OpGenericPtrMemSemantics`. But that's for OpenCL. For Vulkan, which supports `Shader` capability, the spec [requires](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_shadervalidation_a_validation_rules_for_shader_a_href_capability_capabilities_a) "All `<id>` used for Scope and Memory Semantics must be of an OpConstant." So we should be fine here to embed this constant directly into the instruction as an attribute and provide better assembly sugar for it. So +1 to what Mehesh said. If later somebody would like to add support for OpenCL, we can then revisit this. :)

Generally we are missing a principled approach to these "bit enums". I'm thinking introducing something like `BitEnumAttr`. Similar to `Int32EnumAttr`, it describes each possible bit using a `BitEnumAttrCase`. Also similar to `EnumGen`, we can have a TableGen backend to generate conversion functions for them. It should generate conversion functions that take in a int32_t and returns a vector containing the names for each set bit, and vice versa. With that, it would be nice to have something like

```mlir
spv.ControlBarrier "Workgroup", "WorkGroup", "Acquire|UniformMemory"
```

or similar.

And this can be used for other SPIR-V enums like FunctionControl, etc. 

WDYT?

MaheshRavishankar commented on include/mlir/Dialect/SPIRV/SPIRVBase.td:
```diff
@@ -227,6 +230,20 @@ class SPV_Optional<Type type> : Variadic<type>;
+// scope, it specifies the distance of synchronization from the current
+// invocation. When used as an execution scope, it specifies the set of
+// executing invocations taking part in the operation.
+def SPV_Scope : IntOfWidths<[32]>;
```
ot sure about this approach. Should really be using an enum to describe all the valid semantics.. The enum would be an I32Attr (see other enum definitions in SPIRVBase.td). You could then let the operation take a list of these attributes (i.e. and ArrayAttr) and then having the instructions take an array of these

MaheshRavishankar commented on include/mlir/Dialect/SPIRV/SPIRVOps.td:
```diff
@@ -166,6 +166,68 @@ def SPV_CompositeExtractOp : SPV_Op<"CompositeExtract", [NoSideEffect]> {
+  }];
+
+  let arguments = (ins
+    SPV_Scope:$executionScope,
```
If the scope and memory semantics are attributes, then these would attributes as well.

antiagainst commented:
Hey @denis0x0D, FYI, BitEnumAttr is available now after https://github.com/tensorflow/mlir/commit/f317907227b727e6c811f6a1e5d09336486b15ae. :)

denis0x0D commented:
@antiagainst thanks a lot!

denis0x0D commented:
@antiagainst can you please comment on following?
Regarding to spec the "Scope" type is integer that contains bit mask.
https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#Scope_-id-
In this case we need some mechanism to disable autogen for AttrName definition in BitEnumAttr, for example for OpControlBarrier the same type of bit mask "Scope" used for the first operand and for the second. https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#OpControlBarrier
Auto gen generates by default "scope", so we get "redifinition error".
In other case, I was looking on examples in DirextXShaderCompiler and have not seen the uses-cases for that type with multiple bits enabled.
Can you please suggest, what is the better way, to make it as BitEnumAttr and update autogen utils or to use it as IntEnumAttr.
Thanks a lot!

denis0x0D commented:
@antiagainst @MaheshRavishankar I've updated the patch based on BitEnums, rebased on master and force pushed since it completely different version. Can you please take a look? Thanks a lot!

joker-eph commented:
That looks really nice! 
Nicer than the original one with the opaque constant and the SSA value which is still in the PR description (can you update the description by the way?)

denis0x0D commented:
@joker-eph thanks a lot for looking on this! Updated!

MaheshRavishankar commented on lib/Dialect/SPIRV/SPIRVOps.cpp:
```diff
@@ -227,6 +227,33 @@ static LogicalResult verifyMemoryAccessAttribute(LoadStoreOpTy loadStoreOp) {
+        "expected at least one constraint for the memory order");
+  }
+
+  if (bitCount > 1) {
```
I am not sure of whether this is part of SPIR-V spec itself. It might be for specific use cases of SPIR-V, but if its not part of the spec itself, better to not have this validation.

denis0x0D replied:
@MaheshRavishankar thanks a lot for review, I was looking to this link: https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_memory_semantics__id_a_memory_semantics_lt_id_gt

> Despite being a mask and allowing multiple bits to be combined, it is invalid for more than one of these four bits to be set: Acquire, Release, AcquireRelease, or SequentiallyConsistent. Requesting both Acquire and Release semantics is done by setting the AcquireRelease bit, not by setting two bits.

> Memory semantics define memory-order constraints, and on what storage classes those constraints apply to. The memory order constrains the allowed orders in which memory operations in this invocation can made visible to another invocation. The storage classes specify to which subsets of memory these constraints are to be applied. Storage classes not selected are not being constrained.

Do I miss something?

antiagainst replied:
Yeah, this is from the spec.

joker-eph replied:
I think it would be nice to quote and refer to the spec everywhere appropriate as code comment in general. Some part of the clang parser are doing this and this is quite helpful reading code.

denis0x0D replied:
@joker-eph I'll add the comment with link to the spec.

MaheshRavishankar commented on lib/Dialect/SPIRV/Serialization/Serializer.cpp:
```diff
@@ -1523,6 +1553,29 @@ Serializer::processOp<spirv::ExecutionModeOp>(spirv::ExecutionModeOp op) {
+  auto memorySemanticsIntAttr = op.getAttrOfType<IntegerAttr>(
+      spirv::attributeName<spirv::MemorySemantics>());
+
+  auto memoryScopeID = prepareConstantInt(loc, memoryScopeIntAttr);
```
Same here. Maybe the processing of the memoryScope and memorySemantics can be commoned out?

denis0x0D replied:
@MaheshRavishankar thanks for mention this! I'll update it!

denis0x0D commented on lib/Dialect/SPIRV/Serialization/Deserializer.cpp:
```diff
@@ -1910,6 +1965,46 @@ Deserializer::processOp<spirv::FunctionCallOp>(ArrayRef<uint32_t> operands) {
+    return emitError(unknownLoc, "expected 32 bit integer constant for memory "
+                                 "semantics <id> ")
+           << memorySemanticsID << " in OpMemoryBarrier";
+  }
```
@MaheshRavishankar thanks, I'll update it.

antiagainst commented on include/mlir/Dialect/SPIRV/SPIRVBase.td:
```diff
@@ -1034,6 +1037,56 @@ def SPV_MemoryModelAttr :
+  let cppNamespace = "::mlir::spirv";
+}
+
+def SPV_S_CrossDevice : I32EnumAttrCase<"CrossDevice", 0x0000>;
```
Did you use the `utils/define_enum.sh` script to add these fields automatically from the spec? It seems strange here. IIRC, for `I32EnumAttr`, the value here should just be decimal values.

antiagainst commented on include/mlir/Dialect/SPIRV/SPIRVBase.td:
```diff
@@ -1034,6 +1037,56 @@ def SPV_MemoryModelAttr :
+      SPV_MS_ImageMemory
+    ]> {
+  let returnType = "::mlir::spirv::MemorySemantics";
+  let convertFromStorage = "static_cast<::mlir::spirv::MemorySemantics>($_self.getInt())";
```
This is useful to generate getter method for a attribute of this kind.

antiagainst commented on include/mlir/Dialect/SPIRV/SPIRVOps.td:
```diff
@@ -172,6 +172,78 @@ def SPV_CompositeExtractOp : SPV_Op<"CompositeExtract", [NoSideEffect]> {
+    scope ::= `"CrossDevice"` | `"Device"` | `"Workgroup"` | `"Subgroup"`
+              | `"Invocation"`
+
+    memory-semantics ::= `"None"` | `"Acquire"` | `"Release"`
```
No need to list all of them (since this list will keep growing potentially). We can just list a few and then use `| ...` to omit the rest.

antiagainst commented on include/mlir/Dialect/SPIRV/SPIRVOps.td:
```diff
@@ -172,6 +172,78 @@ def SPV_CompositeExtractOp : SPV_Op<"CompositeExtract", [NoSideEffect]> {
+
+  let results = (outs);
+
+  let verifier = [{ return verifyMemorySemantics(*this); }];
```
I think this is just the default and can be removed?

denis0x0D replied:
@antiagainst can you please comment, do I get you right? I have to make `verifyMemorySemantics` code to be autogenerated?

antiagainst replied:
Nvm, this is fine.  Previously I meant that if we delete this line here, because of the SPV_Op definition, we will need to supply a `verify()` method for `spv::ControlBarrierOp` anyway. But that means `verify()` needs to  wrap `verifyMemorySemantics()` here.

antiagainst commented on include/mlir/Dialect/SPIRV/SPIRVOps.td:
```diff
@@ -172,6 +172,78 @@ def SPV_CompositeExtractOp : SPV_Op<"CompositeExtract", [NoSideEffect]> {
+  let verifier = [{ return verifyMemorySemantics(*this); }];
+
+  let autogenSerialization = 0;
+
```
Redundant empty line.

antiagainst commented on include/mlir/Dialect/SPIRV/SPIRVOps.td:
```diff
@@ -266,6 +338,59 @@ def SPV_LoadOp : SPV_Op<"Load", []> {
+    scope ::= `"CrossDevice"` | `"Device"` | `"Workgroup"` | `"Subgroup"`
+               | `"Invocation"`
+
+    memory-semantics ::= `"None"` | `"Acquire"` | `"Release"`
```
Same here.

antiagainst commented on include/mlir/Dialect/SPIRV/SPIRVOps.td:
```diff
@@ -266,6 +338,59 @@ def SPV_LoadOp : SPV_Op<"Load", []> {
+
+  let results = (outs);
+
+  let verifier = [{ return verifyMemorySemantics(*this); }];
```
Same here.

antiagainst commented on include/mlir/Dialect/SPIRV/SPIRVOps.td:
```diff
@@ -266,6 +338,59 @@ def SPV_LoadOp : SPV_Op<"Load", []> {
+  let verifier = [{ return verifyMemorySemantics(*this); }];
+
+  let autogenSerialization = 0;
+
```
Same here.

antiagainst commented on lib/Dialect/SPIRV/SPIRVOps.cpp:
```diff
@@ -227,6 +227,33 @@ static LogicalResult verifyMemoryAccessAttribute(LoadStoreOpTy loadStoreOp) {
+        spirv::bitEnumContains(memorySemantics, memorySemanticsBit));
+  }
+
+  if (!bitCount) {
```
Is this correct? What is the spec rule for this? I think we can have `None`/`Relaxed` memory semantics?

denis0x0D replied:
@antiagainst oh, I missed that none is relaxed, sorry about it. Thanks for mention this!

antiagainst commented on lib/Dialect/SPIRV/SPIRVOps.cpp:
```diff
@@ -227,6 +227,33 @@ static LogicalResult verifyMemoryAccessAttribute(LoadStoreOpTy loadStoreOp) {
+      spirv::MemorySemantics::AcquireRelease,
+      spirv::MemorySemantics::SequentiallyConsistent};
+
+  uint32_t bitCount = 0;
```
I think we can do ```
auto atMostOneInSet = spirv::MemorySemantics::... | ...;

auto bitCount = llvm::countPopulation(memorySemantics & atMostOneInSet);
```

We need to generate `operator&` for bit enums to do the above; but it should be fairly straightforward: https://github.com/tensorflow/mlir/blob/master/tools/mlir-tblgen/SPIRVUtilsGen.cpp#L707

denis0x0D replied:
@antiagainst thanks, I'll update it!

antiagainst commented:
Sorry for the delay! Regarding https://github.com/tensorflow/mlir/pull/130#issuecomment-532372947, I feel I don't quite understand here:

> In this case we need some mechanism to disable autogen for AttrName definition in BitEnumAttr, for example for OpControlBarrier the same type of bit mask "Scope" used for the first operand and for the second. https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#OpControlBarrier
Auto gen generates by default "scope", so we get "redifinition error".

Do you mean the auto-gen'd `attributeName`? I think it should be fine? Could you elaborate a bit?

antiagainst commented:
I think we should just have one `ScopeAttr` definition and fix the autogen tools if they are causing problems. :)

denis0x0D commented:
@antiagainst I was wrong about a special mechanism to disable autogen for attr name, sorry about it. I was relying on parseEnumAttribute function https://github.com/tensorflow/mlir/blob/master/lib/Dialect/SPIRV/SPIRVOps.cpp#L96 , it uses attr name generated by default for Enums https://github.com/tensorflow/mlir/blob/master/tools/mlir-tblgen/SPIRVUtilsGen.cpp#L660
, in case SPV_ScopeAttr, the name is `scope` and if we have the arguments like this:

```
let arguments = (ins
    SPV_ScopeAttr:$execution_scope,
    SPV_ScopeAttr:$memory_scope,
    SPV_MemorySemanticsAttr:$memory_semantics
  );
```
it parses and sets the attribute with default name which is `scope` and not  `execution_scope` or `memory_scope`, so just need to update a little bit a parseEnumAttribute function to allow to parse attribute with custom name.
Thanks.

denis0x0D commented:
@antiagainst @MaheshRavishankar updated regarding to the comments, can you please take a look?
Thanks!

antiagainst commented on lib/Dialect/SPIRV/SPIRVOps.cpp:
```diff
@@ -229,23 +231,20 @@ static LogicalResult verifyMemoryAccessAttribute(LoadStoreOpTy loadStoreOp) {
 
 template <typename BarrierOp>
 static LogicalResult verifyMemorySemantics(BarrierOp op) {
+  // Regarding to the SPIR-V specification:
```
// According to ...
// "..."

We use `` for quoting code.

antiagainst commented on lib/Dialect/SPIRV/Serialization/Deserializer.cpp:
```diff
@@ -132,6 +132,9 @@ class Deserializer {
   /// Gets the constant's attribute and type associated with the given <id>.
   Optional<std::pair<Attribute, Type>> getConstant(uint32_t id);
 
+  /// Gets the constants's integer attribute with the given <id>.
```
/// ... Returns a null IntegerAttr if the given <id> is not registered or does not correspond to an integer constant.

antiagainst commented on lib/Dialect/SPIRV/Serialization/Deserializer.cpp:
```diff
@@ -1875,49 +1886,20 @@ Deserializer::processOp<spirv::ControlBarrierOp>(ArrayRef<uint32_t> operands) {
-    return emitError(unknownLoc, "expected 32 bit integer constant for memory "
-                                 "semantics <id> ")
-           << memorySemanticsID << " in OpControlBarrier";
+  SmallVector<IntegerAttr, 3> argAttrs;
```
Much nicer! Thanks!

denis0x0D commented:
@antiagainst thanks for review! Updated.
